torch/csrc/jit/topological_index.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <c10/util/Exception.h>
+#include <map>
+#include <unordered_map>
+
+namespace torch {
+namespace jit {
+
+template <typename T>
+struct TopologicalIndex;
+
+struct Node;
+using node_topological_index = TopologicalIndex<Node*>;
+using topo_position_t = uint64_t;
+
+// Index to track a topological ordering of nodes. Owned by a block,
+// representing the nodes in the block.
+//
+// This lets us answer questions like "is this node before another node"
+// efficiently, which is useful for optimization. It should be kept up to date
+// with node insertions/deletions by the owning block.
+//
+// The basic scheme is: nodes are assigned topological indices within a
+// 64-bit space. Appending a node moves assigns a position that's a big interval
+// higher than the last node, giving room for insertions in between. If we ever
+// run out of room, we rebuild the index.
+//
+// NOTE: this relies on some implementation details of the node listing, so
+// don't reuse without checking those assumptions.
+template <typename T>
+struct TopologicalIndex {
+ private:
+  // The node list is implemented as a circular linked list, with the output
+  // node as the head/tail. Therefore, the index needs to treat "appends" to the
+  // output as appends to the input.
+  T input_;
+  T output_;
+
+  // Lower and upper bounds of the index. Inclusive range.
+  const topo_position_t lowerBound_;
+  const topo_position_t upperBound_;
+
+  // How far away to space nodes that are appended to the graph.
+  // should be 2^n, where:
+  //   - n is the maximum number of repeated insertions without a re-index
+  //   - 2^(64-n) is the maximum number of appends to the end without reindex
```

Comment:
It will support (as default configured) 2^24 insertions to the end in constant time without needing to rebuild the index. I feel like that's enough? I will take a look at how an interval tree compares.